### PR TITLE
Implement content transmission on AlexandriaNetworkAPI

### PR DESCRIPTION
**Pull request by pipermerriam**
replaces #165 

## What was wrong?

We need high level APIs for retrieving content in Alexandria

## How was it fixed?

Implemented `AlexandriaNetworkAPI.get_content_proof` and `AlexandriaNetworkAPI.get_content_from_nodes`

#### Cute Animal Picture

raccoon-nasty-face

### CODE REVIEW

**Comment by marcgarreau** [comment]
Where does this stand after https://github.com/ethereum/ddht/pull/215?

**Comment by pipermerriam** [comment]
This has been rebased and should be ready for review

File: tests/core/v5_1/alexandria/test_alexandria_network.py [hunk 1]
```diff
@@ -102,3 +105,126 @@ async def test_alexandria_network_responds_to_find_nodes(
...
+
+
+@pytest.mark.trio
+async def test_alexandria_network_get_content_proof_api(
+    alice, bob, alice_alexandria_network, bob_alexandria_client,
+):
+    content = ContentFactory(length=16 * 32 * 4)
+    proof = compute_proof(content, sedes=content_sedes)
+
+    async with bob_alexandria_client.subscribe(GetContentMessage) as subscription:
```
Comment (marcgarreau) [inline tests/core/v5_1/alexandria/test_alexandria_network.py:117]: `GetContentMessage` handles for `get_content_from_nodes` and `get_content_proof`? What's the thinking behind not having a separate proof message type?

Comment (pipermerriam) [reply tests/core/v5_1/alexandria/test_alexandria_network.py:117]: `get_content_proof` is the mid-level primative for retrieving a proof for some content from a single node.

`get_content_from_nodes` is used to retrieve the full content.  It assumes you have figured out all the nodes that have the content you want.  Internally it uses `get_content_proof` to retrieve the content in pieces, re-assembling it locally, and returning the full content once it has been fully retrieved.

Comment (pipermerriam) [reply tests/core/v5_1/alexandria/test_alexandria_network.py:117]: Oh, as for the separation of not having something like `RawContent` and `ContentProof` message types...  It's just really simple to have a single boolean flag and use the same message type.

We also don't have infrastructure for a request that has two different response types.

And we also can't know ahead of time if the content we're requesting will fit inside a single message or need to be split across multiple messages in a proof...

File: ddht/v5_1/alexandria/network.py [hunk 2]
```diff
@@ -153,6 +161,242 @@ async def find_nodes(
...
+            start_chunk_index=start_chunk_index,
+            max_chunks=max_chunks,
+            request_id=request_id,
+        )
+        if response.payload.is_proof:
+            proof = Proof.deserialize(
+                data=response.payload.payload, sedes=content_sedes,
+            )
+        else:
+            proof = compute_proof(
```
Comment (marcgarreau) [inline ddht/v5_1/alexandria/network.py:191]: fyi this logic fork is not yet triggered in tests or benchmarks

**Review by marcgarreau — APPROVED** [review]

File: ddht/v5_1/alexandria/network.py (hunk 2 above)
Comment (pipermerriam) [reply ddht/v5_1/alexandria/network.py:191]: As measured by.... did you run code coverage or something?

Comment (pipermerriam) [reply ddht/v5_1/alexandria/network.py:191]: Ah, nevermind, I see it.

Comment (marcgarreau) [reply ddht/v5_1/alexandria/network.py:191]: Lower-tech :) while manually stepping through stuff, a log/debug was never executed

**Comment by pipermerriam** [comment]
test failure is unrelated.

**Pull request merged**